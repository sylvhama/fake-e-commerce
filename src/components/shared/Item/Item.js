import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import SquareImg from '../SquareImg';
import Like from '../../../svg/like.svg';

const Article = styled.article`
  position: relative;
  background-color: ${props => props.theme.color.white};
  display: flex;
  flex-direction: column;
  overflow: hidden;
`;

const Sold = styled.span`
  position: absolute;
  top: -50px;
  left: -50px;
  z-index: 2;
  width: 100px;
  height: 100px;
  display: flex;
  justify-content: center;
  align-items: flex-end;
  background-color: ${props => props.theme.color.primary};
  color: ${props => props.theme.color.white};
  line-height: 1.5;
  font-size: ${props => props.theme.font.p};
  font-weight: bold;
  text-transform: uppercase;
  letter-spacing: 2px;
  transform: rotate(-45deg);
`;

const Info = styled.div`
  padding: ${props => props.theme.spacing.sm};
  display: flex;
  flex-direction: column;
  flex: 1;
  word-break: break-all;
  &,
  & h2 {
    font-size: ${props => props.theme.font.p};
    color: ${props => props.theme.color.grey};
  }
  & h2 {
    line-height: 1.5;
  }
`;

const Flex = styled.div`
  margin-top: auto;
  display: flex;
  flex-wrap: wrap;
`;

const Likes = styled.span`
  margin-left: auto;
  padding-left: ${props => props.theme.spacing.sm};
  display: flex;
  align-items: center;

  & svg {
    margin-right: 4px;
  }
`;

class Item extends PureComponent {
  render() {
    const { name, price, image, like_count, is_sold_out, ...rest } = this.props;
    return (
      <Article {...rest}>
        {is_sold_out && <Sold>Sold</Sold>}
        <SquareImg src={image} alt={name} />
        <Info>
          <h2>{name}</h2>
          <Flex>
            <strong>Â¥{price.toLocaleString('ja-JP')}</strong>
            <Likes>
              {!!like_count && (
                <Fragment>
                  <Like />
                  <span>{like_count}</span>
                </Fragment>
              )}
            </Likes>
          </Flex>
        </Info>
      </Article>
    );
  }
}

Item.propTypes = {
  name: PropTypes.string,
  price: PropTypes.number,
  image: PropTypes.string,
  like_count: PropTypes.number,
  is_sold_out: PropTypes.bool
};

export default Item;
