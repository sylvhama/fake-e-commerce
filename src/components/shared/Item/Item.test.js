import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../tests/helpers';

import Item from './Item';

test('Item renders correctly', () => {
  const rendered = renderWithTheme(
    <Item name="name" price={1230} image="url" />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});

test('Item renders correctly with likes', () => {
  const rendered = renderWithTheme(
    <Item name="name" price={1230} image="url" like_count={42} />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});

test('Item renders correctly when sold out', () => {
  const rendered = renderWithTheme(
    <Item name="name" price={1230} image="url" is_sold_out={true} />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});
