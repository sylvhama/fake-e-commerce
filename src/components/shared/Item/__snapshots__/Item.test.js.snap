// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders correctly 1`] = `
.c1 {
  position: relative;
  margin: 0;
  padding-top: 100%;
}

.c1 img {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.c0 {
  position: relative;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c2 {
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  word-break: break-all;
}

.c2,
.c2 h2 {
  font-size: 1rem;
  color: #333;
}

.c2 h2 {
  line-height: 1.5;
}

.c3 {
  margin-top: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c4 {
  margin-left: auto;
  padding-left: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 svg {
  margin-right: 4px;
}

<article
  className="c0"
>
  <figure
    className="c1"
  >
    <img
      alt="name"
      src="url"
    />
  </figure>
  <div
    className="c2"
  >
    <h2>
      name
    </h2>
    <div
      className="c3"
    >
      <strong>
        ¥
        1,230
      </strong>
      <span
        className="c4"
      />
    </div>
  </div>
</article>
`;

exports[`Item renders correctly when sold out 1`] = `
.c2 {
  position: relative;
  margin: 0;
  padding-top: 100%;
}

.c2 img {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.c0 {
  position: relative;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c1 {
  position: absolute;
  top: -50px;
  left: -50px;
  z-index: 2;
  width: 100px;
  height: 100px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  background-color: #ea352d;
  color: #fff;
  line-height: 1.5;
  font-size: 1rem;
  font-weight: bold;
  text-transform: uppercase;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
}

.c3 {
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  word-break: break-all;
}

.c3,
.c3 h2 {
  font-size: 1rem;
  color: #333;
}

.c3 h2 {
  line-height: 1.5;
}

.c4 {
  margin-top: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c5 {
  margin-left: auto;
  padding-left: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 svg {
  margin-right: 4px;
}

<article
  className="c0"
>
  <span
    className="c1"
  >
    Sold
  </span>
  <figure
    className="c2"
  >
    <img
      alt="name"
      src="url"
    />
  </figure>
  <div
    className="c3"
  >
    <h2>
      name
    </h2>
    <div
      className="c4"
    >
      <strong>
        ¥
        1,230
      </strong>
      <span
        className="c5"
      />
    </div>
  </div>
</article>
`;

exports[`Item renders correctly with likes 1`] = `
.c1 {
  position: relative;
  margin: 0;
  padding-top: 100%;
}

.c1 img {
  position: absolute;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.c0 {
  position: relative;
  background-color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
}

.c2 {
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  word-break: break-all;
}

.c2,
.c2 h2 {
  font-size: 1rem;
  color: #333;
}

.c2 h2 {
  line-height: 1.5;
}

.c3 {
  margin-top: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c4 {
  margin-left: auto;
  padding-left: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 svg {
  margin-right: 4px;
}

<article
  className="c0"
>
  <figure
    className="c1"
  >
    <img
      alt="name"
      src="url"
    />
  </figure>
  <div
    className="c2"
  >
    <h2>
      name
    </h2>
    <div
      className="c3"
    >
      <strong>
        ¥
        1,230
      </strong>
      <span
        className="c4"
      >
        <IconMock />
        <span>
          42
        </span>
      </span>
    </div>
  </div>
</article>
`;
